docs/adr/adr-013-automatic-gas-price-estimation.md

```diff
@@ -0,0 +1,66 @@
+# ADR #013: Automatic Gas Price Estimation
+
+## Changelog
+
+- 17.01.2024: Initial design
+
+## Status
+
+Proposed
+
+## Context
+
+[CIP 18](https://github.com/celestiaorg/CIPs/blob/main/cips/cip-18.md) defines an interface that a light node can use for transaction submission to retrieve an estimate of the gas price and amount of gas required to include their transaction on Celestia. This predominantly is to allow for greater dynamism during times of congestions where gas prices become greater than the global minimum.
+
+This ADR defines how the go light node will interact with the service as a client. Note this does not cover how the server will actually estimate the gas prices and gas required.
+
+## Decision
+
+Allow users of Celestia nodes to opt in to automatic gas price and gas usage estimation for every write method.
+
+## Detailed Design
+
+### When to estimate the gas price
+
+Users can opt in to automatic price estimation by using the existing `TxConfig.isGasPriceSet`. Previously, if this was false, the `DefaultGasPrice` was used, now if a user isn't manually setting the gas price, the node will query it from the `GasEstimator` service.
+
+### When to estimate the gas required
+
+Depending on the message types submitted, the node will decide whether to also query for the amount of gas that is required to submit the transaction. In the case of the following messages, the gas can be reliably estimated using a simple linear model:
+
+- `MsgPayForBlob`
+
+For all other messages (or combination of messages), the node will call the `GasEstimator` service.
+
+### Setting guardrails
+
+A new field in the `TxConfig` will be introduced allowing for a user to  use the gas price estimation but setting a maximum gas price as a cap that the user is willing to pay. If the price returned by the estimator is greater than the cap, likely due to a price spike, an error will be returned informing the user that the gas price required is currently greater than the user is willing to pay. If the user doesn't specify a cap, a hardcoded value that is 100x the default min gas price will be used (i.e. 0.2 utia).
```

Comment:
Can we assume that [tx priority ](https://github.com/celestiaorg/celestia-app/blob/0e4b98433b084a9a1d05bc8bceec7d6f8b4619e3/proto/celestia/core/v1/gas_estimation/gas_estimator.proto#L36)is medium in case the user wants us to use gas estimator?

Reply:
or we can configure the node during the startup to use some priority by default.

Reply:
Q: why do we need to have any limits if user haven't specified them?

Reply:
The default is medium priority - users don't have to specify it if they don't want to. I guess it should also be present in the Node API as part of the config

Reply:
> Q: why do we need to have any limits if the user hasn't specified them?

Out of an abundance of caution (i.e. limit the damage this can do if the GasEstimation service is faulty)

Reply:
I think it is worth adding that we should not use queried gas price that is less than minGasPrice